File: day4/part2.php
<?php
$input = fopen("input.txt", "r", true) or die("Unable to open file");
$arr = [];

while (!feof($input)) {
    $tempStr = fgets($input);
    $tempArr = str_split(trim($tempStr));

    array_push($arr, $tempArr);
}

$arrUpLeft = $arr;
for($i = 0; $i<count($arr); $i++) {
    $arrUpLeft[$i] = array_pad($arrUpLeft[$i], -$i-count($arr), " ");
    $arrUpLeft[$i] = array_pad($arrUpLeft[$i], count($arr)*2-1,0);
}

$arrUpRight = $arr;
for($i = 0; $i<count($arr); $i++) {
    $arrUpRight[count($arr)-$i-1] = array_pad($arrUpRight[count($arr)-$i-1], -$i-count($arr), " ");
}

echo count(array_intersect(getPos($arrUpLeft), getPos($arrUpRight)));
fclose($input);

function getPos($arr): array {
    $retArr = [];

    $arrT= array_map(null, ...$arr);
    $counter = 0;

    foreach ($arrT as $tempArr) {
        $tempStr = implode("", $tempArr);

        $offset = 0;
        while(strpos($tempStr, "MAS", $offset) !== false) {
            $offset = strpos($tempStr, "MAS", $offset) + 1;
            array_push($retArr,
                ($counter - substr_count(implode("", $arr[$offset]), " ")) . ", $offset"
            );
        }

        $offset = 0;
        while(strpos($tempStr, "SAM", $offset) !== false) {
            $offset = strpos($tempStr, "SAM", $offset) + 1;
            array_push($retArr,
                ($counter - substr_count(implode("", $arr[$offset]), " ")) . ", $offset"
            );
        }

        $counter++;
    }

    return $retArr;
}

File: python/day4/part2.py
def main():
    arr = []

    count = 0
    with open("./day4.txt", "r", ) as file:
        for line in file:
            line = line.strip()
            arr.append(line)

    for i in range(1, len(arr)-1):
        for j in range(1, len(arr[i])-1):
            if arr[i][j] == 'A':
                if not ((arr[i-1][j-1] == 'M' and arr[i+1][j+1] == 'S') or (arr[i-1][j-1] == 'S' and arr[i+1][j+1] == 'M')):
                    continue

                if not ((arr[i-1][j+1] == 'M' and arr[i+1][j-1] == 'S') or (arr[i-1][j+1] == 'S' and arr[i+1][j-1] == 'M')):
                    continue
                
                count += 1
    
    print(count)

if __name__ == "__main__":
    main()